lib/src/vector_math/matrix4.dart

```diff
@@ -660,97 +669,144 @@ class Matrix4 {
-      ty = x.y;
-      tz = x.z;
+      translateByVector4(x);
     } else if (x is double) {
-      tx = x;
-      ty = y;
-      tz = z;
+      translateByDouble(x, y, z, 1.0);
     } else {
       throw UnimplementedError();
     }
+  }
+
+  /// Translate this matrix by x, y, z, w.
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
+  @pragma('dart2js:prefer-inline')
+  void translateByDouble(double tx, double ty, double tz, double tw) {
     final t1 = _m4storage[0] * tx +
         _m4storage[4] * ty +
         _m4storage[8] * tz +
         _m4storage[12] * tw;
+    _m4storage[12] = t1;
+
     final t2 = _m4storage[1] * tx +
         _m4storage[5] * ty +
         _m4storage[9] * tz +
         _m4storage[13] * tw;
+    _m4storage[13] = t2;
+
     final t3 = _m4storage[2] * tx +
         _m4storage[6] * ty +
         _m4storage[10] * tz +
         _m4storage[14] * tw;
+    _m4storage[14] = t3;
+
     final t4 = _m4storage[3] * tx +
         _m4storage[7] * ty +
         _m4storage[11] * tz +
         _m4storage[15] * tw;
-    _m4storage[12] = t1;
-    _m4storage[13] = t2;
-    _m4storage[14] = t3;
     _m4storage[15] = t4;
   }
 
+  /// Translate this matrix by a [Vector3].
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
+  @pragma('dart2js:prefer-inline')
+  void translateByVector3(Vector3 v3) =>
+      translateByDouble(v3.x, v3.y, v3.z, 1.0);
+
+  /// Translate this matrix by a [Vector4].
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
+  @pragma('dart2js:prefer-inline')
+  void translateByVector4(Vector4 v4) =>
+      translateByDouble(v4.x, v4.y, v4.z, v4.w);
+
   /// Multiply this by a translation from the left.
-  /// The translation can be specified with a  [Vector3], [Vector4], or x, y, z.
+  ///
+  /// The translation can be specified with a [Vector3], [Vector4], or x, y, z
+  /// as [double]s.
+  ///
+  /// If you know the argument types in a call site, prefer
+  /// [leftTranslateByDouble], [leftTranslateByVector3], or
+  /// [leftTranslateByVector4] for performance.
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
+  @pragma('dart2js:prefer-inline')
+  @Deprecated('Use leftTranslateByVector3, leftTranslateByVector4, '
+      'or leftTranslateByDouble instead')
   void leftTranslate(dynamic x, [double y = 0.0, double z = 0.0]) {
-    double tx;
-    double ty;
-    double tz;
-    final tw = x is Vector4 ? x.w : 1.0;
     if (x is Vector3) {
-      tx = x.x;
-      ty = x.y;
-      tz = x.z;
+      leftTranslateByVector3(x);
     } else if (x is Vector4) {
-      tx = x.x;
-      ty = x.y;
-      tz = x.z;
+      leftTranslateByVector4(x);
     } else if (x is double) {
-      tx = x;
-      ty = y;
-      tz = z;
+      leftTranslateByDouble(x, y, z, 1.0);
     } else {
       throw UnimplementedError();
     }
+  }
 
+  /// Multiply this by a translation from the left.
+  @pragma('wasm:prefer-inline')
+  @pragma('vm:prefer-inline')
+  @pragma('dart2js:prefer-inline')
```

Comment:
I get inlining the redirecting functions. But these are BIG functions. Should we really be inlining them?

Reply:
@osa1 – wondering if we should hold off on the 2nd level of inlining until we know it's a net win...maybe?

CC @rakudrama

Reply:
This could be counterproductive for JavaScript. I'd rather we didn't prefer-inline for dart2js.

JavaScript has its own JIT, and that can do inlining itself.
If this function is not inlined by dart2js, and super hot, it will be JIT-optimized.
Then all the call sites benefit, even if most of them are not hot.

Over-inlining can cause the host function less likely to be JIT-ed, even if it is hot.
Either is is simply too big to to attempt JIT optimization, or there are more opportunities for de-optimization, and eventually the JIT-optimizer declines to make further attempts.

Any function in JavaScript with more than a few operations is at risk of these counter-productive behaviors.

Reply:
See https://github.com/google/vector_math.dart/pull/347/